rr.go

```diff
@@ -30,32 +34,55 @@ const NameCollision = "127.0.53.53"
 
 // String returns a string representation of an RR in zone-file format.
 func (rr *RR) String() string {
-	return rr.Name + "\t      3600\tIN\t" + rr.Type + "\t" + rr.Value
+	if rr.Ttl == nil {
+		return rr.Name + "\t      3600\tIN\t" + rr.Type + "\t" + rr.Value
+	} else {
+		ttl := ttlString(rr.Ttl)
+		return rr.Name + "\t" + ttl + "\t" + rr.Type + "\t" + rr.Value
+	}
+}
+
+// ttlString constructs the TTL field of an RR string.
+func ttlString(ttl *time.Duration) string {
+	seconds := int(ttl.Seconds())
+	return fmt.Sprintf("%10d", seconds)
 }
 
 // convertRR converts a dns.RR to an RR.
 // If the RR is not a type that this package uses,
 // It will attempt to translate this if there are enough parameters
 // Should all translation fail, it returns an undefined RR and false.
-func convertRR(drr dns.RR) (RR, bool) {
+func convertRR(drr dns.RR, expire bool) (RR, bool) {
+	var ttl *time.Duration
+	var expiry *time.Time
+	if expire {
+		ttl, expiry = calculateExpiry(drr)
+	}
 	switch t := drr.(type) {
 	case *dns.SOA:
-		return RR{toLowerFQDN(t.Hdr.Name), "SOA", toLowerFQDN(t.Ns)}, true
+		return RR{toLowerFQDN(t.Hdr.Name), "SOA", toLowerFQDN(t.Ns), ttl, expiry}, true
 	case *dns.NS:
-		return RR{toLowerFQDN(t.Hdr.Name), "NS", toLowerFQDN(t.Ns)}, true
+		return RR{toLowerFQDN(t.Hdr.Name), "NS", toLowerFQDN(t.Ns), ttl, expiry}, true
 	case *dns.CNAME:
-		return RR{toLowerFQDN(t.Hdr.Name), "CNAME", toLowerFQDN(t.Target)}, true
+		return RR{toLowerFQDN(t.Hdr.Name), "CNAME", toLowerFQDN(t.Target), ttl, expiry}, true
 	case *dns.A:
-		return RR{toLowerFQDN(t.Hdr.Name), "A", t.A.String()}, true
+		return RR{toLowerFQDN(t.Hdr.Name), "A", t.A.String(), ttl, expiry}, true
 	case *dns.AAAA:
-		return RR{toLowerFQDN(t.Hdr.Name), "AAAA", t.AAAA.String()}, true
+		return RR{toLowerFQDN(t.Hdr.Name), "AAAA", t.AAAA.String(), ttl, expiry}, true
 	case *dns.TXT:
-		return RR{toLowerFQDN(t.Hdr.Name), "TXT", strings.Join(t.Txt, "\t")}, true
+		return RR{toLowerFQDN(t.Hdr.Name), "TXT", strings.Join(t.Txt, "\t"), ttl, expiry}, true
 	default:
 		fields := strings.Fields(drr.String())
 		if len(fields) >= 4 {
-			return RR{toLowerFQDN(fields[0]), fields[3], strings.Join(fields[4:], "\t")}, true
+			return RR{toLowerFQDN(fields[0]), fields[3], strings.Join(fields[4:], "\t"), ttl, expiry}, true
 		}
 	}
 	return RR{}, false
 }
+
+// calculateExpiry calculates the expiry time of an RR.
+func calculateExpiry(drr dns.RR) (*time.Duration, *time.Time) {
```

Comment:
Why does this return pointers rather than structs? Seems this would increase GC pressure due to the additional heap allocations. `time.Duration` is only 64 bits, same as a pointer.

Reply:
And `time.Time` should be passed by value, it's also just two 64-bit fields.

Reply:
Didn't even realize, I was thinking pointers are more efficient.

Tried updating everything to use values. Instead of checking for `nil` it checks for equal to `emptyTime`. Does that make sense?